### PR TITLE
add version metric

### PR DESCRIPTION
**Pull request by WilliButz**
I just saw that dnsmasq does not only provide the metrics mentioned on the manpage.
Apart from those, it also exposes [the following](http://thekelleys.org.uk/gitweb/?p=dnsmasq.git;a=blob;f=src/option.c;h=65df93aff291a55d6f9b56a356654afb744ea62f;hb=HEAD#l4698) under the bind domain:
```
version == "dnsmasq-2.79"
authors == "Simon Kelley"
copyright == "Copyright (c) 2000-2018 Simon Kelley"
```
While I can't see any reason for the exporter to expose author and copyright information, I think the version number is quite useful, so this PR only adds the version metric.

### CODE REVIEW

File: dnsmasq.go [hunk 1]
```diff
@@ -150,8 +157,22 @@ func (s *server) metrics(w http.ResponseWriter, r *http.Request) {
...
+						return err
+					} else {
+						metricValue = s[1]
+					}
+				} else {
+					metricValue = txt.Txt[0]
+				}
+				f, err := strconv.ParseFloat(metricValue, 64)
 				if err != nil {
+					log.Warnln("could not parse metric value:", err)
```
Comment (stapelberg) [inline dnsmasq.go]: Move the log.Warnln call to line 205 please, so that all errors are covered.

File: dnsmasq.go [hunk 2]
```diff
@@ -150,8 +157,22 @@ func (s *server) metrics(w http.ResponseWriter, r *http.Request) {
 				if got, want := len(txt.Txt), 1; got != want {
 					return fmt.Errorf("stats DNS record %q: unexpected number of replies: got %d, want %d", txt.Hdr.Name, got, want)
 				}
-				f, err := strconv.ParseFloat(txt.Txt[0], 64)
+				var metricValue string
+				// splits version string 'dnsmasq-<version>'
+				if s := strings.Split(txt.Txt[0], "-"); txt.Hdr.Name == "version.bind." {
```
Comment (stapelberg) [inline dnsmasq.go]: This is needlessly expensive: we don’t need to split all values, just the value when the name is version.bind.

How about:
```go
metricValue := txt.Txt[0]
if txt.Hdr.Name == "version.bind." && strings.HasPrefix(txt.Txt[0], "dnsmasq-") {
  metricValue = strings.TrimPrefix(txt.Txt[0], "dnsmasq-")
}
```

**Comment by WilliButz** [comment]
@stapelberg I think it is just nice to have.
One could for example use it to compare the running dnsmasq version to the latest release and get alerted if there is a difference.
Of course this could also be done by alerting on new tags in the repo's atom feed or similar :)

Regarding the changing version numbers, I guess there wouldn't be any difference to a non-changing value because the values all belong to the same timeseries in prometheus as there are no different labels to distinguish them by.

File: dnsmasq.go (hunk 1 above)
Comment (WilliButz) [reply dnsmasq.go]: I also removed the other calls of log.Warnln which I added before, as they are now also covered :)

**Comment by stapelberg** [comment]
Thanks for the update, the code looks much better now.

Given that software has to be maintained, and maintenance gets more complicated the more code we have to support, I suggest we shelve this PR for when a specific and strong need comes along: for the time being, I’m not convinced that having the version available as a metric is really that useful. As you confirmed, there are other ways to keep software up-to-date, which I imagine work much better.

Thanks for understanding,

**Pull request closed without merging**